### PR TITLE
plugin-ext: validate path when unpacking archives

### PR DESCRIPTION
**Pull request by marechal-p**
#### What it does

Fix zip-slip by validating where a given file will be unpacked. If the
expected path is outside of the destination folder: log a warning and
ignore the file.

Fixes https://github.com/eclipse-theia/theia/issues/7319

#### How to test

Tests should fail if you comment lines 32 to 44 in `plugin-deployer-file-handler-context-impl.ts`.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/plugin-ext/src/main/node/plugin-deployer-file-handler-context-impl.ts [hunk 1]
```diff
@@ -24,8 +26,30 @@ export class PluginDeployerFileHandlerContextImpl implements PluginDeployerFileH
...
+                if (!expectedFilePath.startsWith(path.join(absoluteDestPath, path.sep))) {
+                    zipSlipFiles.add([file.path, expectedFilePath]);
+                    return false; // only skip the exploit files, maybe the rest is fine.
+                } else {
+                    return true;
+                }
+            }
+        });
+        if (zipSlipFiles.size > 0) {
+            console.error(`Detected a zip-slip exploit in archive: "${sourcePath}"`);
```
Comment (akosyakov) [inline packages/plugin-ext/src/main/node/plugin-deployer-file-handler-context-impl.ts]: Should not we rather throw?

**Comment by marechal-p** [comment]
TBH I don't really care. I like `decompress` because it handles several extensions at the time. The vulnerability only happens when you let the library unpack without validating anything on the user side, which this PR adds.

I'll still update this PR based on Anton's comment, feel free to open a separate PR that would replace the dependency by something else.

File: packages/plugin-ext/src/main/node/plugin-deployer-file-handler-context-impl.ts (hunk 1 above)
Comment (marechal-p) [reply packages/plugin-ext/src/main/node/plugin-deployer-file-handler-context-impl.ts]: Now throwing.

File: packages/plugin-ext/package.json [hunk 2]
```diff
@@ -81,7 +81,9 @@
     "@types/escape-html": "^0.0.20",
     "@types/lodash.clonedeep": "^4.5.3",
     "@types/ps-tree": "^1.1.0",
-    "@types/request": "^2.0.3"
+    "@types/request": "^2.0.3",
+    "chai": "^4.2.0",
```
Comment (akosyakov) [inline packages/plugin-ext/package.json]: shared dev dependencies should be managed in one place in the root package

**Comment by marcdumais-work** [comment]
> Also, if someone still happens to use `decompress` in another area of the code, it will not benefit from the same patching.

I find this to be a compelling argument. If we keep `decompress` it will always show-up as a known vulnerability for the platform and applications that use `plugin-ext`. We'll need to explain that we took steps so that the vulnerability is not triggered for our use-case, and that it can be safely ignored. After a little while we will just ignore it, knowing that "it's ok in this case", which can lead to problems if the library gets used in other places without taking these protective steps.

**Comment by marechal-p** [comment]
Preferred way is to change the dependency.

**Pull request closed without merging**